clang/test/CXX/basic/basic.lookup/basic.lookup.classref/p1-cxx11.cpp

```diff
@@ -55,15 +55,21 @@ namespace PR11856 {
 
   template<typename T> T *end(T*);
 
-  class X { };
+  struct X { };
+  struct Y {
+    int end;
+  };
   template <typename T>
   void Foo2() {
     T it1;
-    if (it1->end < it1->end) {
-    }
+    if (it1->end < it1->end) { }
 
     X *x;
-    if (x->end < 7) {  // expected-error{{no member named 'end' in 'PR11856::X'}}
-    }
+    if (x->end < 7) { } // expected-error{{expected '>'}}
+                        // expected-note@-1{{to match this '<'}}
+                        // expected-error@-2{{expected unqualified-id}}
```

Comment:
This doesn't look expected to me, what is going on?

Reply:
This actually is expected -- see the final note in the comments of the PR. Since we didn't find anything via qualified lookup, we "lock into" the nested-name-specifier interpretation (because it would otherwise be an invalid class member access since such a member doesn't exist). So, we do unqualified lookup, find the template, and try parse `<` as a template argument list.

Reply:
I see.

Still, error recovery is not great here.

I think this looks sufficiently different from a template argument list that we should just error out and commit to the interpretation that this is a comparison.

We already have this capability to look ahead and tell if something looks like a template argument list, see: https://github.com/llvm/llvm-project/blob/d07362f7a9fc06e2445f5c4bc62c10a339bf68a5/clang/include/clang/Parse/Parser.h#L2734

Reply:
@mizvekov I implemented this in the ~5 most recent commits, if you'd like to take a second look

Reply:
Yes, this looks great now, thanks!

I still have to finish review of the rest of this huge patch, this is going to take a while :)